src/AIS.cpp

```diff
@@ -129,3 +136,147 @@ ais_base::VoyageInformation AIS::getVoyageInformation(ais::message_05 const& mes
+    }
+    else if (!std::isnan(course_over_ground.getRad()) &&
+             speed_over_ground >= MIN_SPEED_FOR_VALID_COURSE) {
+        return {Eigen::Quaterniond(Eigen::AngleAxisd(course_over_ground.getRad(),
+                    Eigen::Vector3d::UnitZ())),
+            ais_base::PositionCorrectionStatus::POSITION_CENTERED_USING_COURSE};
+    }
+    else {
+        return {Eigen::Quaterniond::Identity(),
+            ais_base::PositionCorrectionStatus::POSITION_RAW};
+    }
+}
+
+/**
+ * @brief Transforms the sensor position relative to the vessel into the world frame
+ *
+ * @param sensor2vessel_pos The position of the sensor relative to the vessel
+ * @param vessel2world_ori The vessel's orientation in the world frame
+ *
+ * @return The sensor's position in the world frame
+ */
+base::Vector3d sensorToVesselInWorldPose(base::Vector3d const& sensor2vessel_pos,
+    Eigen::Quaterniond const& vessel2world_ori)
+{
+    base::Vector3d sensor2vessel_in_world_pos;
+    sensor2vessel_in_world_pos = vessel2world_ori * sensor2vessel_pos;
+
+    return sensor2vessel_in_world_pos;
+}
+
+/**
+ * @brief Converts a GPS position into UTM coordinates
+ *
+ * @param position The GPS position
+ * @param utm_converter The UTM converter
+ *
+ * @return A RigidBodyState in UTM coordinates
+ */
+base::samples::RigidBodyState convertGPSToUTM(ais_base::Position const& position,
+    gps_base::UTMConverter const& utm_converter)
+{
+    gps_base::Solution sensor2world_solution;
+    sensor2world_solution.latitude = position.latitude.getDeg();
+    sensor2world_solution.longitude = position.longitude.getDeg();
+
+    base::samples::RigidBodyState sensor2world_UTM;
+    sensor2world_UTM.position =
+        utm_converter.convertToUTM(sensor2world_solution).position;
+    sensor2world_UTM.time = base::Time::now();
+
+    return sensor2world_UTM;
+}
+
+/**
+ * @brief Converts a UTM position back to GPS coordinates in the world frame
+ *
+ * @param sensor2world_UTM The sensor position in UTM coordinates
+ * @param sensor2vessel_in_world_pos The sensor position in the world frame
+ * @param utm_converter The UTM converter
+ *
+ * @return A pair containing the latitude and longitude
+ */
+std::pair<base::Angle, base::Angle> convertUTMToGPSInWorldFrame(
+    base::samples::RigidBodyState const& sensor2world_UTM,
+    base::Vector3d const& sensor2vessel_in_world_pos,
+    gps_base::UTMConverter const& utm_converter)
+{
+    base::samples::RigidBodyState vessel2world_UTM;
+    vessel2world_UTM.position = sensor2world_UTM.position + sensor2vessel_in_world_pos;
+
+    gps_base::Solution vessel2world_GPS;
+    vessel2world_GPS = utm_converter.convertUTMToGPS(vessel2world_UTM);
+
+    return {base::Angle::fromDeg(vessel2world_GPS.latitude),
+        base::Angle::fromDeg(vessel2world_GPS.longitude)};
+}
+
+ais_base::Position AIS::applyPositionCorrection(
+    std::optional<ais_base::Position> position,
```

Comment:
I'm missing the logic for this `std::optional`. You are throwing if the optional is not present, meaning that the value is *not* optional.

Just a const reference. If there's a place where the value is indeed optional, its presence needs to be checked by the caller.